### PR TITLE
Add 3d media type / mesh support

### PR DESCRIPTION
**Pull request by sashankaryal**
## Changes
- Added a new media type 3d to support 3D visualizations.
- Implemented classes for `Scene`, `Material`, `Mesh`, `Shape`, `PointCloud`, and `Light`, along with their respective serialization and deserialization logic.
- Add rendering logic tailored for FiftyOne 3d scenes.
- Add support for exporting/importing fo3d files

## Testing
- Unit tests for newly introduced SDK classes
- Currently working on Playwright tests to cover end-to-end workflows

## Summary by CodeRabbit

- **New Features**
	- Added support for 3D datasets in the application, including new functionalities for 3D visualization.
	- Introduced configuration updates for language settings and review workflows.
	- Enhanced the app with chat auto-reply functionality.
	- Added installation of specific dependencies for building documentation.
	- Comprehensive updates to support rendering of both "3d" and "point_cloud" media types, including new components for 3D visualization.
	- Updated documentation to reflect support for 3D datasets and new visualization functionalities.
- **Refactor**
	- Refactored the configuration loading process for more flexibility in plugin loading.
	- Modified functionality related to 3D slices visibility and introduced new logic for conditional rendering based on 3D slice presence.
	- Updated import paths and condition checks to improve consistency and support new 3D functionalities.
- **Bug Fixes**
	- Fixed issues related to 3D media type handling in GraphQL queries and state management.
- **Tests**
	- Updated end-to-end tests to reflect changes in attribute checks and support for new 3D functionalities.
- **Chores**
	- Enhanced utility functions and introduced new test cases for improved functionality.

### CODE REVIEW

File: requirements/common.txt [hunk 1]
```diff
@@ -19,6 +19,7 @@ plotly==5.17.0
 pprintpp==0.4.0
 psutil>=5.7.0
 pymongo>=3.12
+pydantic==2.5.3
```
Comment (benjaminpkane) [inline requirements/common.txt:22]: Minimum `pydantic` req will need to be determined and added to `setup.py`

File: tests/unittests/threed/object_3d_tests.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
+"""
+FiftyOne sample-related unit tests.
+
+| Copyright 2017-2023, Voxel51, Inc.
```
Comment (benjaminpkane) [inline tests/unittests/threed/object_3d_tests.py:4]:
```suggestion
| Copyright 2017-2024, Voxel51, Inc.
```

**Comment by sashankaryal** [comment]
@coderabbitai review

File: app/packages/looker-3d/src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    !Object.hasOwn(scene, "visible") ||
+    !Object.hasOwn(scene, "position") ||
+    !Object.hasOwn(scene, "quaternion") ||
+    !Object.hasOwn(scene, "scale") ||
+    !Object.hasOwn(scene, "children") ||
+    !Array.isArray(scene.children)
+  ) {
+    throw new InvalidSceneError();
+  }
+
```
[the remark this answers is not in the document]
Comment (sashankaryal) [reply app/packages/looker-3d/src/utils.ts:103]: Yes

File: docs/source/user_guide/using_datasets.rst [hunk 4]
```diff
@@ -4333,10 +4333,195 @@ To get started exploring video datasets, try loading the
...
+    scene = fo.Scene()
+    sphere1 = fo.SphereGeometry("sphere1", radius=2.0)
+    sphere1.position = [-1, 0, 0]
+    sphere1.default_material.color = "red"
+
+    sphere2 = fo.SphereGeometry("sphere2", radius=2.0)
+    sphere2.position = [-1, 0, 0]
+    sphere2.default_material.color = "blue"
+
+    mesh = fo.GLTFMesh("mesh", filepath="mesh.glb")
```
Comment (benjaminpkane) [inline docs/source/user_guide/using_datasets.rst:4378]: Getting an argument error here
```python
----> 9 mesh = fo.GLTFMesh("mesh", filepath="mesh.glb")
     10 mesh.rotation = fo.Euler(90, 0, 0, degrees=True)
     11 scene.add(sphere1, sphere2, mesh)

TypeError: __init__() missing 1 required positional argument: 'gltf_path'
```

File: docs/source/user_guide/using_datasets.rst [hunk 5]
```diff
@@ -4333,10 +4333,195 @@ To get started exploring video datasets, try loading the
...
+`PCD <https://pointclouds.org/documentation/tutorials/pcd_file_format.html>`_
+point cloud format. A code snippet to create a PCD object that can be added
+to a Fiftyone 3D scene is shown below:
+
+.. code-block:: python
+    :linenos:
+
+    import fiftyone as fo
+
+    fo_pcd = fo.Pointcloud("my-pcd",
```
Comment (benjaminpkane) [inline docs/source/user_guide/using_datasets.rst:4432]: Via `black`
```suggestion
    fo_pcd = fo.Pointcloud(
        "my-pcd", "/path/to/point-cloud.pcd", flag_for_projection=True
    )
```

File: app/packages/looker-3d/src/fo3d/AssetErrorBoundary.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  static getDerivedStateFromError(error) {
+    console.error(error);
+    return { hasError: true };
+  }
+
+  render() {
+    // @ts-ignore
+    if (this.state.hasError) {
+      // todo: add indicator in canvas that asset failed loading
+      return null;
```
Comment (benjaminpkane) [inline app/packages/looker-3d/src/fo3d/AssetErrorBoundary.tsx:18]: Can in indicator / informative message be added? Is separate `AssetErrorBoundary` required?

File: app/packages/looker-3d/src/fo3d/AssetWrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import { useCursor } from "@react-three/drei";
+import { useCallback, useState } from "react";
+import { useSetRecoilState } from "recoil";
+import { FoSceneNode } from "../hooks";
+import { activeNodeAtom, isStatusBarOnAtom } from "../state";
+
+// note: raycaster in point cloud (Points.raycast()) is exceptionally slow and causes significant performance issues, so we disable it
+// todo: use a custom raycaster to improve performance
```
Comment (benjaminpkane) [inline app/packages/looker-3d/src/fo3d/AssetWrapper.tsx:8]: Proposing we track this internally and remove the unused component for now

File: app/packages/looker-3d/src/fo3d/FoScene.tsx [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        <R3fNode key={child.name} node={child} visibilityMap={visibilityMap} />
+      );
+    });
+  }, [node, visibilityMap]);
+
+  const label = useMemo(() => getLabelForSceneNode(node), [node]);
+
+  const isNodeVisible = useMemo(
+    () => Boolean(visibilityMap[label]),
+    [visibilityMap, node]
```
Comment (benjaminpkane) [inline app/packages/looker-3d/src/fo3d/FoScene.tsx:217]: This is missing the `label` dependency. Is that ok?

File: app/packages/looker-3d/src/fo3d/Gizmos.tsx [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+      Math.abs(sceneBoundingBox.max.x) === Infinity ||
+      !upVector
+    ) {
+      return [100, 100];
+    }
+
+    let maxInOrthoNormalPlane: number;
+
+    // account for the possibility that the scene is not centered at the origin
+    let offset: number;
```
Comment (benjaminpkane) [inline app/packages/looker-3d/src/fo3d/Gizmos.tsx:85]: `offset` is unused 🤔

**Review by benjaminpkane — APPROVED** [review]
Amazing addition! 🚀  Merged #4202 with some small edits

LGTM! @swheaton may have comments, but from my perspective we can merge follow up in separate PRs to refine

**Review by swheaton — CHANGES_REQUESTED** [review]
taking another pass.

However I would like to see all code rabbit suggestions resolved before merging. If they are irrelevant, comment why (it learns based on this) and resolve the comment.

**Review by swheaton — APPROVED** [review]

**Comment by swheaton** [comment]
Nominal approval - we decided we will merge and further changes requested may be made in a follow up PR

**Pull request merged**